nixos/modules/misc/version.nix

```diff
@@ -44,6 +44,49 @@ let
   };
   initrdRelease = pkgs.writeText "initrd-release" (attrsToText initrdReleaseContents);
 
+  checkRelease = version:
+    let
+      parts = lib.versions.splitVersion version;
+      isVersion = lib.length parts == 2 && lib.all (p: lib.stringLength p == 2) parts;
+      majorVersion = lib.toIntBase10 (lib.elemAt parts 0);
+      minorVersion = lib.elemAt parts 1;
+
+      versionPatterns = [
+        # only 13.10
+        { fromMajor = 13; minor = [ "10" ]; }
+        # 14.04 and 14.12
+        { fromMajor = 14; minor = [ "04" "12" ]; }
+        # only 15.09
+        { fromMajor = 15; minor = [ "09" ]; }
+        # 16.03 to 20.09
+        { fromMajor = 16; minor = [ "03" "09" ]; }
+        # from 21.05
+        { fromMajor = 21; minor = [ "05" "11" ]; }
+      ];
+
+      # find the versioning pattern that applies by looking for the first
+      # major version newer than `majorVersion`, and picking the previous pattern
+      patternIndex = lib.lists.findFirstIndex
+        ({ fromMajor, ... }: fromMajor > majorVersion)
+        (lib.length versionPatterns)
+        versionPatterns;
+
+      validMinorVersions =
+        if patternIndex == 0
+        then []
+        else (lib.elemAt versionPatterns (patternIndex - 1)).minor;
+
+      correctMinorVersion = lib.elem minorVersion validMinorVersions;
+      notNewerThanNixpkgs = lib.versionAtLeast trivial.release version;
```

Comment:
Why do we have to play this game instead of just listing all previous versions?

Reply:
The intention was to avoid adding yet another place that needs to be changed when a new release is cut (just when the release schedule changes). It was neater before I added pre-16 releases with less regularity.

Reply:
`system.stateVersion` was added in d166c854b6ecfd9ada520439b06f7eb9e3e4be7c and its first valid value was actually, uh… 15.07(!). So you can omit the earlier releases but have to account for past schedule drift. Looking at `git log -p .version`, you also need 21.03, 17.10 (though this only existed for about 20 minutes, so you can probably omit it), 15.08, and 15.06. (There were actually also 15.05, 15.04, 14.11 (looks like 14.12 never existed on `master`?), 14.10, and 14.02, but as I said those predate `stateVersion`.)

Reply:
As [previously mentioned](https://github.com/NixOS/nixpkgs/pull/317858#issuecomment-2303087830), there is a `stateVersion` even in [the NixOS infrastructure](https://github.com/NixOS/infra/blob/c8b2e1baf8a562bef4e103ba33ad28c5226e774b/build/haumea/default.nix#L52) using a value from before the option was introduced, so I can't omit it.

As for the `.version` log, that's a good point. I've been mostly going off of https://channels.nixos.org/ for the authoritative list of "real" versions. But that might be unfair to unstable users, if their config was generated while `.version` didn't refer to a channel version.

There are some very early versions with most likely zero instances in the wild, going back to 0.5 in 2004, in that history, but I'll include all of the ones following the YY.MM format (starting at 13.07), if it means we don't have to discuss which of them are "real enough". It'll still benefit users going forward, even if it might be overly-permissive for some (long) past releases. Please leave a thumbs-up or reply if you think we should just include *all of them (YY.MM)*.

Reply:
Sorry for missing the previous discussion. Just going based on the `.version` log as you suggested seems good to me.

Reply:
You can't be expected to read all the previous iterations of this PR :)

It's not just `.version`, as channels.nixos.org contains some versions that have never been in master/`.version`, so I included it if it appeared in either.